sys/net/gnrc/network_layer/sixlowpan/iphc/gnrc_sixlowpan_iphc.c

```diff
@@ -368,7 +369,74 @@ size_t gnrc_sixlowpan_iphc_decode(gnrc_pktsnip_t *ipv6, gnrc_pktsnip_t *pkt, siz
 
     }
 
-    /* TODO: add next header decoding */
+    /* next header decoding */
+    if (iphc_hdr[IPHC1_IDX] & SIXLOWPAN_IPHC1_NH) {
+
+        uint8_t udp_nhc = iphc_hdr[IPHC1_IDX + 2];
+        udp_hdr_t udp_hdr;
+        uint8_t tmp;
+        uint8_t octets_to_move = pkt->size - payload_offset;
+
+        if ((udp_nhc  & SIXLOWPAN_NHC_UDP_MASK) != SIXLOWPAN_NHC_UDP_ID) {
+            DEBUG("6lo iphc nhc: unspecified header type\n");
+            return 0;
+        }
+
+        if ((udp_nhc & SIXLOWPAN_NHC_UDP_C_MASK) == SIXLOWPAN_NHC_UDP_C_ELIDED) {
+            DEBUG("6lo iphc nhc: unsupported elided checksum\n");
+            return 0;
+        }
+
+        ipv6_hdr->nh = PROTNUM_UDP;
+        payload_offset += 1;
+        octets_to_move -= 1;
+
+        switch (udp_nhc & SIXLOWPAN_NHC_UDP_PP_MASK) {
+
+            case SIXLOWPAN_NHC_UDP_SD_INLINE:
+                /* TODO */
+                memcpy(&(udp_hdr.src_port), &iphc_hdr[IPHC1_IDX + 3], 2);
+                memcpy(&udp_hdr.dst_port, &iphc_hdr[IPHC1_IDX + 5], 2);
+                octets_to_move -= 4;
+                break;
+
+            case SIXLOWPAN_NHC_UDP_D_INLINE:
+                /* TODO */
+                memcpy(&udp_hdr.src_port, &iphc_hdr[IPHC1_IDX + 3], 2);
+                udp_hdr.dst_port = byteorder_htons(iphc_hdr[IPHC1_IDX + 4] + 0xf000);
+                octets_to_move -= 3;
+                break;
+
+            case SIXLOWPAN_NHC_UDP_S_INLINE:
+                /* TODO */
+                udp_hdr.src_port = byteorder_htons(iphc_hdr[IPHC1_IDX + 3] + 0xf000);
+                memcpy(&udp_hdr.dst_port, &iphc_hdr[IPHC1_IDX + 4], 2);
+                octets_to_move -= 3;
+                break;
+
+            case SIXLOWPAN_NHC_UDP_SD_ELIDED:
+                /* TODO */
+                tmp = iphc_hdr[IPHC1_IDX + 3];
+                udp_hdr.src_port = byteorder_htons((tmp >> 4) + 0xf0b0);
+                udp_hdr.src_port = byteorder_htons((tmp & 0xf) + 0xf0b0);
+                octets_to_move -= 1;
+                break;
+
+            default:
+                break;
+        }
+
+        /* make place for udp length */
+        if(gnrc_pktbuf_realloc_data(pkt, pkt->size + 2)) {
+            printf("realloc failed\n");
```

Comment:
I assume this is just for debugging purpose while implementing. But should be replaced with `DEBUG` and a proper cleanup `pktbuf_free(pkt); return 0;`